HIP/hip-0000-atomic-association-contracts.md

```diff
@@ -0,0 +1,104 @@
+
+## Specification
+
+Hedera Smart Contract Service optimistically allows custody of any token for smart contracts, and checks at the end of a contract call that tokens are associated to entities whose balance is changing from zero to nonzero. 
+
+### Successful
+
+The response code for token transfer of a token not associated to a smart contract, whose balance is equal to zero at the end of the transaction, is SUCCESS = 22.
+
+### Revert
+
+The response code for token transfer of a token not associated to a smart contract, whose balance is nonzero at the end of the transaction, is TOKEN_NOT_ASSOCIATED_TO_ACCOUNT = 184, or a new enum TOKEN_NOT_ASSOCIATED_TO_CONTRACT.
+
+## User stories
+
+As a solidity developer, I do not want to associate tokens to a smart contract that custodies tokens but is not designed to have token balances.
+
+## Backwards Compatibility
+
+This would be backward compatible.
+
+## Security Implications
+
+Increased complexity of Hedera Smart Contract Service may lead to vulnerabilities.
+
+## How to Teach This
+
+N/A
+
+## Reference Implementation
+#### Example :
+Example of a transfer transaction that should be allowed:
+
+```
+/**
+     * @dev Example function to transfer entire balance of unassociated tokens
+     * @param token The HTS token to transfer
+     * @param treasury The recipient of the token transfer (is associated to token)
+     */
+     function function1(address token, address recipient) external {
+        require(IERC20(token).balanceOf(address(this)) == 0);
+        HederaTokenService.dissociateToken(token, address(this));
+        // assumes msg.sender approved an allowance to address(this)
+        HederaTokenService.transferToken(token, msg.sender, address(this), 1);
+        int respCode = HederaTokenService.transferToken(token, address(this), recipient, 1);
+     }
+```
+Example of a transfer transaction that should not be allowed:
+
+```
+/**
+     * @dev Example of a function call that should fail because address(this) balance is nonzero at end of function call
+     * @param token The HTS token to transfer
+     * @param treasury The recipient of the token transfer (is associated to token)
+     */
+     function function2(address token, address recipient) external {
+        require(IERC20(token).balanceOf(address(this)) == 0);
+        HederaTokenService.dissociateToken(token, address(this));
+        // assumes msg.sender approved an allowance to address(this)
+        HederaTokenService.transferToken(token, msg.sender, address(this), 2);
+        HederaTokenService.transferToken(token, address(this), recipient, 1);
+     }
+ ```
+
+## Rejected Ideas
+
+None as of yet.
+
+## Open Issues
+
+None as of yet.
+
+## References
+
+None.
```

Comment:
I'd recommend adding references to:
- EIP-20
- HTS
- anything specific to HTS associateToken an dissociateToken
- anything specific to response codes (e.g. where `TOKEN_NOT_ASSOCIATED_TO_ACCOUNT` and `SUCCESS` are defined)